src/safemode_ui.cpp

```diff
@@ -0,0 +1,740 @@
+                draw_column( COLUMN_ATTITUDE, Creature::get_attitude_ui_data( rule.attitude ).first );
+                draw_column( COLUMN_PROXIMITY, ( !rule.exclude ) ? to_string( rule.proximity ).c_str() : "---" );
+                draw_column( COLUMN_INCLUDE_EXCLUDE, ( rule.exclude ) ?  _( "Exclude" ) : _( "Include" ) );
+            }
+        }
+
+        wrefresh( w );
+
+        const std::string action = ctxt.handle_input();
+
+        if( action == "NEXT_TAB" ) {
+            tab++;
+            if( tab >= MAX_TAB ) {
+                tab = 0;
+                line = 0;
+            }
+        } else if( action == "PREV_TAB" ) {
+            tab--;
+            if( tab < 0 ) {
+                tab = MAX_TAB - 1;
+                line = 0;
+            }
+        } else if( action == "QUIT" ) {
+            break;
+        } else if( tab == CHARACTER_TAB && g->u.name.empty() ) {
+            //Only allow loaded games to use the char sheet
+        } else if( action == "DOWN" ) {
+            line++;
+            if( line >= ( int )current_tab.size() ) {
+                line = 0;
+            }
+        } else if( action == "UP" ) {
+            line--;
+            if( line < 0 ) {
+                line = current_tab.size() - 1;
+            }
+        } else if( action == "ADD_RULE" ) {
+            changes_made = true;
+            current_tab.push_back( rules_class( "", true, false, Creature::A_HOSTILE,
+                                                get_option<int>( "SAFEMODEPROXIMITY" ) ) );
+            line = current_tab.size() - 1;
+        } else if( action == "REMOVE_RULE" && current_tab_non_empty ) {
+            changes_made = true;
+            current_tab.erase( current_tab.begin() + line );
+            if( line > ( int )current_tab.size() - 1 ) {
+                line--;
+            }
+            if( line < 0 ) {
+                line = 0;
+            }
+        } else if( action == "COPY_RULE" && current_tab_non_empty ) {
+            changes_made = true;
+            current_tab.push_back( rules_class(
+                                       current_tab[line].rule,
+                                       current_tab[line].active,
+                                       current_tab[line].exclude,
+                                       current_tab[line].attitude,
+                                       current_tab[line].proximity ) );
+            line = current_tab.size() - 1;
+        } else if( action == "SWAP_RULE_GLOBAL_CHAR" && current_tab_non_empty ) {
+            if( ( tab == GLOBAL_TAB && g->u.name != "" ) || tab == CHARACTER_TAB ) {
+                changes_made = true;
+                //copy over
+                auto &temp_rules_from = ( tab == GLOBAL_TAB ) ? global_rules : character_rules;
+                auto &temp_rules_to = ( tab == GLOBAL_TAB ) ? character_rules : global_rules;
+
+                temp_rules_to.push_back( rules_class(
```

Comment:
Looks like you're copying the `rules_class` object. Why not use the copy constructor for that?
```C++
temp_rules_to.push_back(temp_rules_from[...]);
```